### PR TITLE
Example ADPUPA bufr YAMLs

### PR DESCRIPTION
**Pull request by PraveenKumar-NOAA**
## Description

This PR adds a number of YAMLs that convert ADPUPA data from bufr to IODA-formatted netCDF using the _bufr2ioda_x_.
The following files are new:

**YAMLs**

- test/testinput/bufr_ncep_adpupa.yaml
- test/testinput/bufr_ncep_adpupa_nc002001.yaml
- test/testinput/bufr_ncep_adpupa_nc002002.yaml
- test/testinput/bufr_ncep_adpupa_nc002003.yaml
- test/testinput/bufr_ncep_adpupa_nc002004.yaml
- test/testinput/bufr_ncep_adpupa_nc002005.yaml

**INPUT BUFRs**

- test/testinput/gdas.t12z.adpupa.tm00.bufr_d
- test/testinput/NC002001
- test/testinput/NC002003
- test/testinput/NC002005

**OUTPUT IODA (NCs)**

- test/testoutput/gdas.t12z.adpupa_bufr.tm00.nc
- test/testoutput/gdas.t12z.adpupa_bufr_nc002001.tm00.nc
- test/testoutput/gdas.t12z.adpupa_bufr_nc002003.tm00.nc
- test/testoutput/gdas.t12z.adpupa_bufr_nc002005.tm00.nc

**Note:**

- The test data, 2020110112, only has three subsets NC002001, NC002003, and NC002005 in it.
- A separate YAML, input bufr, and output ioda file is created for each subset.
- Ioda-validation is performed using an ObsSpace.yaml, which was free of errors and warnings.
- A separate PR will be issued for the ObsSpace.yaml, which will add/update a few adpupa variables into this yaml.
- A separate issue is created for the test data validation (Validation of ioda-converter for the adpupa bufr data  #1033), where I created plots for the pressure, temperature, wind, and station elevation.

### Issue(s) addressed

Link the issues to be closed with this PR
-fixes:
#513 
#515
#909
#910 
#911 
#912
#1033

## Acceptance Criteria (Definition of Done)

PR is finished when all YAMLs are accepted and passed the validation tests.

## Dependencies

There are no dependencies.

### CODE REVIEW

File: test/testinput/bufr_ncep_adpupa.yaml
Comment (nicholasesposito) [inline test/testinput/bufr_ncep_adpupa.yaml:34]: Just FYI, I don't think the brackets [] are required since there don't appear to be multiple sequences.

File: test/testinput/bufr_ncep_adpupa_nc002001.yaml
Comment (nicholasesposito) [inline test/testinput/bufr_ncep_adpupa_nc002001.yaml:23]: Does MINU need NC002001?

File: test/testinput/bufr_ncep_adpupa.yaml
Comment (nicholasesposito) [inline test/testinput/bufr_ncep_adpupa.yaml:147]: Consider changing "moisture" to what the variable actually is. Check the source code to figure out what it might be (like dewpoint or relative humidity)

File: test/testinput/bufr_ncep_adpupa.yaml
Comment (nicholasesposito) [inline test/testinput/bufr_ncep_adpupa.yaml:152]: I changed mine to two variables: QualityMarker/windSpeed and windDirection. wind might be too vague.

File: test/testinput/bufr_ncep_adpupa_nc002001.yaml
Comment (nicholasesposito) [inline test/testinput/bufr_ncep_adpupa_nc002001.yaml:62]: Is NC002001 required for this?

File: test/testinput/bufr_ncep_adpupa.yaml
Comment (nicholasesposito) [reply test/testinput/bufr_ncep_adpupa.yaml:147]: I see Dewpoint above so that will probably be what you want to use.

File: test/testinput/bufr_ncep_adpupa_nc002001.yaml
Comment (nicholasesposito) [inline test/testinput/bufr_ncep_adpupa_nc002001.yaml:11]: Are we comfortable leaving the filename input as NC002001, etc?

**Review by nicholasesposito — COMMENTED** [review]
Some general comments and questions about usage of QM names and subset name additions. Once they've been answered, please do the same for the other yamls

File: test/testinput/bufr_ncep_adpupa.yaml
Comment (rmclaren) [inline test/testinput/bufr_ncep_adpupa.yaml:42]: Brackets are not necessary as they only contain 1 query.

**Review by rmclaren — CHANGES_REQUESTED** [review]
Be careful mixing queries like */ and NC00xxxx/ ... You might end up with a lot of missing value rows. If your input files only consist of the one subset anyways, you probably don't need to do this (just make everything a * query).

Alternatively you can specify the list of subsets to use without explicitly mentioning them in every query by adding an exports::subsets field as in the following example:

```yaml
   exports:
        group_by_variable: pressure  
        subsets:
          - NC002001
        variables: ....
```

and then just make all your queries */whatever.

File: test/testinput/bufr_ncep_adpupa.yaml
Comment (PraveenKumar-NOAA) [reply test/testinput/bufr_ncep_adpupa.yaml]: Removed all the [] from the YAMLs, @nicholasesposito thanks for catching this.

File: test/testinput/bufr_ncep_adpupa_nc002001.yaml
Comment (PraveenKumar-NOAA) [reply test/testinput/bufr_ncep_adpupa_nc002001.yaml:23]: NO. The NC002001 subset has no info about MINU variable.

File: test/testinput/bufr_ncep_adpupa.yaml
Comment (PraveenKumar-NOAA) [reply test/testinput/bufr_ncep_adpupa.yaml]: Thanks, the changes have been made by replacing moisture by depointTemperature.

File: test/testinput/bufr_ncep_adpupa.yaml
Comment (PraveenKumar-NOAA) [reply test/testinput/bufr_ncep_adpupa.yaml]: @nicholasesposito wind has been replaced by windSpeed for this QualityMarker. 
I don't think we need windDirectionQM as there is only one variable for the  wind QM.

File: test/testinput/bufr_ncep_adpupa_nc002001.yaml
Comment (PraveenKumar-NOAA) [reply test/testinput/bufr_ncep_adpupa_nc002001.yaml:64]: Has been fixed by using the subset in the export section, as per @rmclaren suggestion.

**Comment by PraveenKumar-NOAA** [comment]
> Some general comments and questions about usage of QM names and subset name additions. Once they've been answered, please do the same for the other yamls

@nicholasesposito I applied the above changes for all the subset YAMLs.

File: test/testinput/bufr_ncep_adpupa.yaml
Comment (PraveenKumar-NOAA) [reply test/testinput/bufr_ncep_adpupa.yaml]: @rmclaren these changes have been implemented and [] were removed.

Comment (PraveenKumar-NOAA) [reply test/testinput/bufr_ncep_adpupa.yaml]: @rmclaren fixed and updated.

File: test/testinput/bufr_ncep_adpupa_nc002003.yaml
Comment (nicholasesposito) [inline test/testinput/bufr_ncep_adpupa_nc002003.yaml:1]: Change the copyrights to 2022

**Review by nicholasesposito — COMMENTED** [review]
Once the copyrights are changed, I'll approve. Otherwise, looks good to me!

File: test/testinput/bufr_ncep_adpupa_nc002003.yaml
Comment (PraveenKumar-NOAA) [reply test/testinput/bufr_ncep_adpupa_nc002003.yaml]: @nicholasesposito changed the year from 2020 to 2022, thanks.

**Comment by PraveenKumar-NOAA** [comment]
> Once the copyrights are changed, I'll approve. Otherwise, looks good to me!

Changed and updated.

**Review by nicholasesposito — APPROVED** [review]

File: test/testinput/bufr_ncep_adpupa.yaml
Comment (PraveenKumar-NOAA) [reply test/testinput/bufr_ncep_adpupa.yaml]: @rmclaren could you review this again?

**Review by rmclaren — APPROVED** [review]

**Review by YoulongXia-NOAA — APPROVED** [review]

**Comment by PraveenKumar-NOAA** [comment]
@emilyhcliu @srherbener please review and merge this PR for the ADPUPA bufr dump YAMLs.

**Review by srherbener — COMMENTED** [review]
I'm not expert at the yaml configuration, but others who are have approved so I think those are okay. However, I did notice that the data files (input bufr, output netcdf) are large for unit testing. If these files are involved with unit (test tier1) testing, then they need to be trimmed down. They have 100,000s of locations, and that number (for tier1 tests) needs to be more around 100 (or less). However, if these files are for a higher tier testing (tier2, tier3 for example), then their size is okay. Note the tier1 (unit) tests are targeting code coverage, whereas the higher level tiers are for scientific validation, system level testing, performance checking, etc.

**Comment by PraveenKumar-NOAA** [comment]
@srherbener I don't know if these files are for tier1 or for the higher ones. @emilyhcliu please let me know if I need to thin these files, for about 100 locations.

**Review by srherbener — APPROVED** [review]
Thanks for reducing the test file sizes. Looks great!

**Comment by PraveenKumar-NOAA** [comment]
@srherbener thank you. Please also merge this branch, as @BenjaminRuston wants this to be done asap.

**Comment by srherbener** [comment]
I can merge - just need to wait for the CI testing to complete (and pass of course).

**Comment by srherbener** [comment]
Something seems wrong with the CI tests - they have been waiting for a long time (haven't even started yet). I'm going to try triggering the tests again with an empty commit.

**Comment by PraveenKumar-NOAA** [comment]
@srherbener great, thanks.

**Pull request merged**
